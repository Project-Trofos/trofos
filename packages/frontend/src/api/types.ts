export type Project = {
  id: number;
  pname: string;
  pkey: string | null;
  description: string | null;
  course_id: number | null;
  public: boolean;
  created_at: string;
  course?: Course;
  is_archive: boolean | null;
};

export type Course = {
  id: number;
  code: string;
  startYear: number;
  startSem: number;
  endYear: number;
  endSem: number;
  cname: string;
  description: string | null;
  public: boolean;
  created_at: string;
  shadow_course: boolean;
  is_archive: boolean | null;
};

export type User = {
  user_email: string;
  user_display_name: string;
  user_id: number;
  projects: Project[];
  courses: CourseRoles[];
  basicRoles: BasicRoles[];
};

export type BasicRoles = {
  user_email: string;
  role_id: number;
};

export type CourseRoles = {
  id: number;
  user_email: string;
  role_id: number;
  course_id: number;
};

export type CreateUserRequest = {
  userEmail: string;
  newPassword: string;
};

export type RegisterUser = {
  userEmail: string;
  newPassword: string;
  userDisplayName: string;
};

export type UserCourseRoleRequest = {
  id: number;
  userRole: number;
  userId: number;
};

export type Role = {
  role_name: string;
  id: number;
};

export type Action = {
  action: string;
};

export type ActionOnRole = {
  id: number;
  action: string;
};

export type ActionsOnRoles = {
  id: number;
  role_name: string;
  actions: Action[];
};

export type ProjectData = Project & {
  course: Course | null;
  users: UserData[];
  sprints: {
    id: number;
    name: string;
  }[];
  backlogStatuses: Omit<BacklogStatusData, 'project_id'>[];
  telegramChannelLink: string;
  owner_id: number | null;
};

export type UserData = {
  user: Pick<User, 'user_id' | 'user_email' | 'user_display_name' | 'courses'>;
};

export type ScrumBoardUserData = {
  user: {
    user_id: number | null;
    user_display_name: string;
    user_email: string;
  };
};

export type Milestone = {
  id: number;
  name: string;
  start_date: string;
  deadline: string;
  course_id: number;
  created_at: string;
};

export type Announcement = {
  id: number;
  title: string;
  content: string;
  user_id: number;
  created_at: string;
  updated_at?: string;
};

// CourseData type returned by BE api. We perform transformResponse to convert the data into type CourseData
export type CourseDataResponse = Omit<CourseData, 'users'> & {
  courseRoles: UserData[];
};

export type CourseData = Course & {
  milestones: Milestone[];
  announcements: Announcement[];
  users: UserData[];
};

export type UserOnRolesOnCourse = {
  id: number;
  user_id: number;
  role_id: number;
  course_id: number;
  role: {
    id: number;
    role_name: string;
  };
};

export type BacklogStatusData = {
  name: string;
  type: 'todo' | 'in_progress' | 'done';
  project_id: number;
  order: number;
};

export type Comment = {
  comment_id: number;
  backlog_id: number;
  project_id: number;
  commenter_id: number;
  content: string;
  created_at: string;
  updated_at: string | null;
  commenter: {
    created_at: string;
    project_id: string;
    user: {
      user_id: number;
      user_email: string;
      user_display_name: string;
    };
  };
};

export type CommentFieldsType = {
  projectId: number;
  backlogId: number;
  commenterId: number;
  content: string;
};

export enum BacklogStatus {
  TODO = 'To do',
  DONE = 'Done',
}

export type BacklogPriority = 'very_high' | 'high' | 'medium' | 'low' | 'very_low' | null;

export type Backlog = {
  backlog_id: number;
  summary: string;
  type: 'story' | 'task' | 'bug';
  priority: BacklogPriority;
  reporter_id: number;
  assignee_id: number | null;
  sprint_id: number | null;
  points: number | null;
  description: string | null;
  project_id: number;
  status: BacklogStatus | string;
  assignee: {
    created_at: string;
    project_id: number;
    user_id: number;
    user: {
      user_email: string;
      user_display_name: string;
    };
  } | null;
  epic_id?: number | null;
  epic?: Epic | null;
};

export type Epic = {
  epic_id: number;
  project_id: number;
  name: string;
  description: string | null;
};

export type DefaultBacklog = {
  backlog_id?: number;
  summary?: string;
  type?: 'story' | 'task' | 'bug';
  priority?: BacklogPriority;
  reporter_id?: number;
  assignee_id?: number | null;
  sprint_id?: number | null;
  points?: number | null;
  description?: string | null;
  project_id?: number;
  status?: BacklogStatus | string;
  assignee?: {
    created_at: string;
    project_id: number;
    user_id: number;
    user: {
      user_email: string;
      user_display_name: string;
    };
  } | null;
  epic_id?: number | null;
};

export type BacklogUpdatePayload = {
  projectId: number;
  backlogId: number;
  srcSprintId?: number | null;
  fieldToUpdate: Partial<Backlog>;
};

export enum BacklogHistoryType {
  CREATE = 'create',
  UPDATE = 'update',
  DELETE = 'delete',
}

export type BacklogHistory = {
  history_type: BacklogHistoryType;
  date: string;
} & Omit<Backlog, 'assignee' | 'summary' | 'description'>;

export type UpdateUserRolePayload = {
  userId: number;
  newRoleId: number;
};

export type ProjectGitLink = {
  project_id: number;
  repo: string;
};

export type ProjectGitLinkData = {
  projectId: number;
  repoLink: string;
};

export type ProjectUserSettings = {
  project_id: number;
  user_id: number;
  email_notification: boolean;
};

export type Settings = {
  current_year: number;
  current_sem: number;
};

export type RetrospectiveVote = {
  id: number;
  retro_id: number;
  user_id: number;
  type: RetrospectiveVoteType;
};

export type Retrospective = {
  id: number;
  sprint_id: number;
  content: string;
  type: RetrospectiveType;
  score: number;
  votes: RetrospectiveVote[];
  is_action_taken: boolean;
};

export enum RetrospectiveType {
  POSITIVE = 'positive',
  NEGATIVE = 'negative',
  ACTION = 'action',
}

export enum RetrospectiveVoteType {
  UP = 'up',
  DOWN = 'down',
}

export type CourseImportCsvPayload = {
  courseId: number;
  payload: FormData;
};

export type Feedback = {
  id: number;
  user_id?: number;
  user: {
    user_email: string;
    user_display_name: string;
  };
  sprint_id: number;
  content: string;
  created_at: string;
  updated_at?: string;
};

export type OAuth2Payload = {
  code: string;
  state: string;
  callbackUrl: string;
};

export type Invite = {
  project_id: number;
  email: string;
  expiry_date: Date;
};

export type UserApiKey = {
  id: number;
  user_id: number;
  api_key: string;
  created_at: string;
  last_used: string | null;
  active: boolean;
};

export type UserGuideQueryResponse = {
  answer: string;
  links: Array<string>;
};

export type FeatureFlag = {
  feature_name: string;
  active: boolean;
};

export type ProjectAssignment = {
  id: number;
  targetProjectId: number;
  sourceProjectId: number;
};
