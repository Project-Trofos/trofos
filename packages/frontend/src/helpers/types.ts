export type PickRename<T, K extends keyof T, R extends PropertyKey> = Pick<Omit<T, K> & { [P in R]: T[K] }, R>;
