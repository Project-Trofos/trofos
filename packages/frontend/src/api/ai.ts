import trofosApiSlice from '.';
import { UserGuideQueryResponse, UserGuideRecommendation } from './types';

const extendedApi = trofosApiSlice.injectEndpoints({
  endpoints: (builder) => ({
    answerUserGuideQuery: builder.mutation<UserGuideQueryResponse, { query: string; isEnableMemory: boolean }>({
      query: ({ query, isEnableMemory }) => ({
        url: `ai/userGuideQuery`,
        credentials: 'include',
        method: 'POST',
        body: { query, isEnableMemory },
      }),
    }),
    recommendUserGuideSections: builder.mutation<UserGuideRecommendation[], void>({
      query: () => ({
        url: `ai/recommendUserGuide`,
        method: 'POST',
        credentials: 'include',
      }),
    }),
  }),
});

export const { useAnswerUserGuideQueryMutation, useRecommendUserGuideSectionsMutation } = extendedApi;
