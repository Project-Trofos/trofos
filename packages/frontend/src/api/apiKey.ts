import trofosApiSlice from '.';
import { UserApiKey } from './types';

const extendedApi = trofosApiSlice.injectEndpoints({
  endpoints: (builder) => ({
    getUserApiKey: builder.query<UserApiKey, void>({
      query: () => ({
        url: 'api-key/me',
        credentials: 'include',
      }),
      providesTags: ['UserApiKey'],
    }),
    generateUserApiKey: builder.mutation<UserApiKey, void>({
      query: () => ({
        url: 'api-key/generate',
        method: 'POST',
        credentials: 'include',
      }),
      invalidatesTags: ['UserApiKey'],
    }),
  }),
});

export const { useGetUserApiKeyQuery, useGenerateUserApiKeyMutation } = extendedApi;
