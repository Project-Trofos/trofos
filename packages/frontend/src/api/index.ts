import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react';

// Define a service using a base URL and expected endpoints
const trofosApiSlice = createApi({
  reducerPath: 'trofosApi',
  baseQuery: fetchBaseQuery({
    baseUrl: process.env.NODE_ENV === 'test' ? 'http://localhost:3001/api' : '/api',
    prepareHeaders: (headers) => {
      headers.set('Cache-control', 'no-cache');
      return headers;
    },
  }),
  tagTypes: [
    'Project',
    'UserInfo',
    'Course',
    'Backlog',
    'BacklogHistory',
    'User',
    'Sprint',
    'ActionsOnRoles',
    'CourseRoles',
    'ProjectRoles',
    'Settings',
    'StandUpHeader',
    'StandUp',
    'Retrospective',
    'ProjectUserSettings',
    'Feedback',
  ],
  endpoints: () => ({}),
});

// Define a service using a base URL and expected endpoints
export const nusmodsApiSlice = createApi({
  reducerPath: 'nusmodsApi',
  baseQuery: fetchBaseQuery({ baseUrl: 'https://api.nusmods.com/v2/2022-2023' }),
  endpoints: () => ({}),
});

export default trofosApiSlice;
