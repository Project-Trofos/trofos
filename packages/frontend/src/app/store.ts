import { configureStore } from '@reduxjs/toolkit';
import { setupListeners } from '@reduxjs/toolkit/query';
import trofosApiSlice, { nusmodsApiSlice } from '../api';

const store = configureStore({
  reducer: {
    [trofosApiSlice.reducerPath]: trofosApiSlice.reducer,
    [nusmodsApiSlice.reducerPath]: nusmodsApiSlice.reducer,
  },
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware().concat(trofosApiSlice.middleware).concat(nusmodsApiSlice.middleware),
});

// optional, but required for refetchOnFocus/refetchOnReconnect behaviors
// see `setupListeners` docs - takes an optional callback as the 2nd arg for customization
setupListeners(store.dispatch);

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;

export default store;
