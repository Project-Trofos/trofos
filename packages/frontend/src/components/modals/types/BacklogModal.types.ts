export interface BacklogFormFields extends FormData {
  projectId: number;
}

export type BacklogSelect = {
  id: string;
  name: string;
};
