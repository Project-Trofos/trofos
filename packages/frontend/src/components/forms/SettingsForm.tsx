import React, { useEffect } from 'react';
import { Form, DatePicker, Select, Button, message } from 'antd';
import dayjs from 'dayjs';
import { getErrorMessage } from '../../helpers/error';
import { Settings } from '../../api/types';
import { useUpdateSettingsMutation } from '../../api/settings';

type SettingsFormPropType = {
  settings: Settings | undefined;
};

const CURRENT_YEAR = 'current-year';
const CURRENT_SEMESTER = 'current-semester';

export default function SettingsForm(props: SettingsFormPropType) {
  const { settings } = props;

  const [form] = Form.useForm();

  const [updateSettings] = useUpdateSettingsMutation();

  const handleSave = async () => {
    try {
      const setting = {
        current_year: form.getFieldsValue()[CURRENT_YEAR].year(),
        current_sem: form.getFieldsValue()[CURRENT_SEMESTER],
      } as Settings;
      await updateSettings(setting).unwrap();
      message.success('System settings successfully updated!');
    } catch (error) {
      message.error(getErrorMessage(error));
    }
  };

  useEffect(() => {
    form.resetFields();
  }, [settings, form]);

  return (
    <>
      <Form
        form={form}
        initialValues={{
          [CURRENT_YEAR]: dayjs().year(settings?.current_year ?? dayjs().year()),
          [CURRENT_SEMESTER]: settings?.current_sem ?? 1,
        }}
      >
        <Form.Item name={CURRENT_YEAR} key={CURRENT_YEAR} rules={[{ required: true }]} label="Current Year">
          <DatePicker picker="year" style={{ width: 200 }} size="middle" />
        </Form.Item>
        <Form.Item name={CURRENT_SEMESTER} key={CURRENT_SEMESTER} rules={[{ required: true }]} label="Current Semester">
          <Select placeholder="Select a semester" style={{ width: 200 }} size="middle">
            <Select.Option key="1" value={1}>
              1
            </Select.Option>
            <Select.Option key="2" value={2}>
              2
            </Select.Option>
          </Select>
        </Form.Item>
      </Form>
      <Button type="primary" onClick={handleSave}>
        Save
      </Button>
    </>
  );
}
