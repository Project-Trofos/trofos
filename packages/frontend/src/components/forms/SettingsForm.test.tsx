import { render, screen } from '@testing-library/react';
import React from 'react';
import { Provider } from 'react-redux';
import { Settings } from '../../api/types';
import store from '../../app/store';
import SettingsForm from './SettingsForm';

describe('SettingsForm test', () => {
  const mockSettings: Settings = {
    current_sem: 1,
    current_year: 2022,
  };

  function renderForm() {
    return render(
      <Provider store={store}>
        <SettingsForm settings={mockSettings} />
      </Provider>,
    );
  }

  describe('when rendering', () => {
    it('should render according to the snapshot', () => {
      const { baseElement } = renderForm();
      expect(baseElement).toMatchSnapshot();
    });
  });
});
