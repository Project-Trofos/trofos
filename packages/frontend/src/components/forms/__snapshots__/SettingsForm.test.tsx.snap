// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SettingsForm test > when rendering > should render according to the snapshot 1`] = `
<body>
  <div>
    <form
      class="ant-form ant-form-horizontal"
    >
      <div
        class="ant-form-item"
      >
        <div
          class="ant-row ant-form-item-row"
        >
          <div
            class="ant-col ant-form-item-label"
          >
            <label
              class="ant-form-item-required"
              for="current-year"
              title="Current Year"
            >
              Current Year
            </label>
          </div>
          <div
            class="ant-col ant-form-item-control"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <div
                  class="ant-picker ant-picker-middle"
                  style="width: 200px;"
                >
                  <div
                    class="ant-picker-input"
                  >
                    <input
                      aria-required="true"
                      autocomplete="off"
                      id="current-year"
                      placeholder="Select year"
                      readonly=""
                      size="12"
                      title="2022"
                      value="2022"
                    />
                    <span
                      class="ant-picker-suffix"
                    >
                      <span
                        aria-label="calendar"
                        class="anticon anticon-calendar"
                        role="img"
                      >
                        <svg
                          aria-hidden="true"
                          data-icon="calendar"
                          fill="currentColor"
                          focusable="false"
                          height="1em"
                          viewBox="64 64 896 896"
                          width="1em"
                        >
                          <path
                            d="M880 184H712v-64c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v64H384v-64c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v64H144c-17.7 0-32 14.3-32 32v664c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V216c0-17.7-14.3-32-32-32zm-40 656H184V460h656v380zM184 392V256h128v48c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-48h256v48c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-48h128v136H184z"
                          />
                        </svg>
                      </span>
                    </span>
                    <span
                      class="ant-picker-clear"
                      role="button"
                    >
                      <span
                        aria-label="close-circle"
                        class="anticon anticon-close-circle"
                        role="img"
                      >
                        <svg
                          aria-hidden="true"
                          data-icon="close-circle"
                          fill="currentColor"
                          focusable="false"
                          height="1em"
                          viewBox="64 64 896 896"
                          width="1em"
                        >
                          <path
                            d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm165.4 618.2l-66-.3L512 563.4l-99.3 118.4-66.1.3c-4.4 0-8-3.5-8-8 0-1.9.7-3.7 1.9-5.2l130.1-155L340.5 359a8.32 8.32 0 01-1.9-5.2c0-4.4 3.6-8 8-8l66.1.3L512 464.6l99.3-118.4 66-.3c4.4 0 8 3.5 8 8 0 1.9-.7 3.7-1.9 5.2L553.5 514l130 155c1.2 1.5 1.9 3.3 1.9 5.2 0 4.4-3.6 8-8 8z"
                          />
                        </svg>
                      </span>
                    </span>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="ant-form-item"
      >
        <div
          class="ant-row ant-form-item-row"
        >
          <div
            class="ant-col ant-form-item-label"
          >
            <label
              class="ant-form-item-required"
              for="current-semester"
              title="Current Semester"
            >
              Current Semester
            </label>
          </div>
          <div
            class="ant-col ant-form-item-control"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <div
                  aria-required="true"
                  class="ant-select ant-select-in-form-item ant-select-single ant-select-show-arrow"
                  style="width: 200px;"
                >
                  <div
                    class="ant-select-selector"
                  >
                    <span
                      class="ant-select-selection-search"
                    >
                      <input
                        aria-activedescendant="current-semester_list_0"
                        aria-autocomplete="list"
                        aria-controls="current-semester_list"
                        aria-haspopup="listbox"
                        aria-owns="current-semester_list"
                        aria-required="true"
                        autocomplete="off"
                        class="ant-select-selection-search-input"
                        id="current-semester"
                        readonly=""
                        role="combobox"
                        style="opacity: 0;"
                        type="search"
                        unselectable="on"
                        value=""
                      />
                    </span>
                    <span
                      class="ant-select-selection-item"
                      title="1"
                    >
                      1
                    </span>
                  </div>
                  <span
                    aria-hidden="true"
                    class="ant-select-arrow"
                    style="user-select: none;"
                    unselectable="on"
                  >
                    <span
                      aria-label="down"
                      class="anticon anticon-down ant-select-suffix"
                      role="img"
                    >
                      <svg
                        aria-hidden="true"
                        data-icon="down"
                        fill="currentColor"
                        focusable="false"
                        height="1em"
                        viewBox="64 64 896 896"
                        width="1em"
                      >
                        <path
                          d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                        />
                      </svg>
                    </span>
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </form>
    <button
      class="ant-btn ant-btn-primary"
      type="button"
    >
      <span>
        Save
      </span>
    </button>
  </div>
</body>
`;
